crates/wasmtime/src/runtime/gc/enabled/structref.rs

```diff
@@ -0,0 +1,757 @@
+///         }
+///         Err(e) => return Err(e),
+///     };
+///
+///     // That instance's field should have the expected value.
+///     let val = my_struct.field(&mut scope, 0)?.unwrap_i32();
+///     assert_eq!(val, 42);
+///
+///     // And we can update the field's value because it is a mutable field.
+///     my_struct.set_field(&mut scope, 0, Val::I32(36))?;
+///     let new_val = my_struct.field(&mut scope, 0)?.unwrap_i32();
+///     assert_eq!(new_val, 36);
+/// }
+/// # Ok(())
+/// # }
+/// # foo().unwrap();
+/// ```
+#[derive(Debug)]
+#[repr(transparent)]
+pub struct StructRef {
+    pub(super) inner: GcRootIndex,
+}
+
+unsafe impl GcRefImpl for StructRef {
+    #[allow(private_interfaces)]
+    fn transmute_ref(index: &GcRootIndex) -> &Self {
+        // Safety: `StructRef` is a newtype of a `GcRootIndex`.
+        let me: &Self = unsafe { mem::transmute(index) };
+
+        // Assert we really are just a newtype of a `GcRootIndex`.
+        assert!(matches!(
+            me,
+            Self {
+                inner: GcRootIndex { .. },
+            }
+        ));
+
+        me
+    }
+}
+
+impl StructRef {
+    /// Allocate a new `struct` and get a reference to it.
+    ///
+    /// # Errors
+    ///
+    /// If the allocation cannot be satisfied because the GC heap is currently
+    /// out of memory, but performing a garbage collection might free up space
+    /// such that retrying the allocation afterwards might succeed, then a
+    /// [`GcHeapOutOfMemory<()>`][crate::GcHeapOutOfMemory] error is returned.
+    ///
+    /// # Panics
+    ///
+    /// Panics if the allocator, or any of the field values, is not associated
+    /// with the given store.
+    pub fn new(
+        mut store: impl AsContextMut,
+        allocator: &StructRefPre,
+        fields: &[Val],
+    ) -> Result<Rooted<StructRef>> {
+        Self::_new(store.as_context_mut().0, allocator, fields)
+    }
+
+    pub(crate) fn _new(
+        store: &mut StoreOpaque,
+        allocator: &StructRefPre,
+        fields: &[Val],
+    ) -> Result<Rooted<StructRef>> {
+        assert_eq!(
+            store.id(),
+            allocator.store_id,
+            "attempted to use a `StructRefPre` with the wrong store"
+        );
+
+        // Type check the given values against the field types.
+        let expected_len = allocator.ty.fields().len();
+        let actual_len = fields.len();
+        ensure!(
+            actual_len == expected_len,
+            "expected {expected_len} fields, got {actual_len}"
+        );
+        for (ty, val) in allocator.ty.fields().zip(fields) {
+            assert!(
+                val.comes_from_same_store(store),
+                "field value comes from the wrong store",
+            );
+            let ty = match ty.element_type() {
+                StorageType::I8 | StorageType::I16 => &ValType::I32,
+                StorageType::ValType(ty) => ty,
+            };
```

Comment:
Oh I thought you made a helper that went from storage type to val type?

Reply:
heh, I wrote the helper only after I'd already open coded it a bunch and I guess I missed this spot

Reply:
ah, actually I wrote the helper for `wasmtime_types::FieldType` not `wasmtime::FieldType`... we might want to actually do some trait nastiness to dedupe all this stuff at some point...

Reply:
oh no nvm again: it is a method on `StorageType` not `FieldType`, d'oh